components/form/form-element-mixin.js

```diff
@@ -0,0 +1,239 @@
+import { getLabelText } from './form-helper.js';
+import { LocalizeCoreElement } from '../../lang/localize-core-element.js';
+
+export class FormElementValidityState {
+
+	constructor(flags) {
+		this.flags = {
+			...{
+				valueMissing: false,
+				typeMismatch: false,
+				patternMismatch: false,
+				tooLong: false,
+				tooShort: false,
+				rangeUnderflow: false,
+				rangeOverflow: false,
+				stepMismatch: false,
+				badInput: false,
+				customError: false
+			},
+			...flags
+		};
+	}
+
+	get badInput() {
+		return this.flags.badInput;
+	}
+
+	get customError() {
+		return this.flags.customError;
+	}
+
+	get patternMismatch() {
+		return this.flags.patternMismatch;
+	}
+
+	get rangeOverflow() {
+		return this.flags.rangeOverflow;
+	}
+
+	get rangeUnderflow() {
+		return this.flags.rangeUnderflow;
+	}
+
+	get stepMismatch() {
+		return this.flags.stepMismatch;
+	}
+
+	get tooLong() {
+		return this.flags.tooLong;
+	}
+
+	get tooShort() {
+		return this.flags.tooShort;
+	}
+
+	get valid() {
+		return Object.values(this.flags).reduce((f1, f2) => f1 && !f2, true);
+	}
+
+	get valueMissing() {
+		return this.flags.valueMissing;
+	}
+
+}
+
+export const FormElementMixin = superclass => class extends LocalizeCoreElement(superclass) {
+
+	constructor() {
+		super();
+		this._validationCustoms = new Set();
+		this._validationMessage = '';
+		this._validity = new FormElementValidityState({});
+		this.formValue = null;
+
+		this.addEventListener('d2l-validation-custom-connected', this._validationCustomConnected);
+		this.addEventListener('d2l-validation-custom-disconnected', this._validationCustomDisconnected);
+	}
+
+	checkValidity() {
+		return this.validity.valid;
+	}
+
+	get formAssociated() {
+		return true;
+	}
+
+	get label() {
+		return null;
+	}
+
+	get labelText() {
+		const label = this.label || getLabelText(this);
+		if (label) {
+			return label;
+		}
+		console.warn(this, ' is missing a label');
+		return this.localize('components.form-element-mixin.defaultFieldLabel');
+	}
+
+	setCustomValidity(message) {
+		this._validity = new FormElementValidityState({ customError: true });
+		this._validationMessage = message;
+	}
+
+	setFormValue(formValue) {
+		this.formValue = formValue;
+	}
+
+	setValidity(flags) {
```

Comment:
Is the expectation that the user passes one or more of the flags indicated in the constructor above?  What happens when they don't pass anything?